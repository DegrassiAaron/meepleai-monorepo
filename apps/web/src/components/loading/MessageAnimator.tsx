/**
 * MessageAnimator Component
 *
 * Wrapper component for animating chat messages with direction-specific slide-in effects.
 * - 'left': AI messages sliding from left (x: -20)
 * - 'right': User messages sliding from right (x: 20)
 *
 * Respects user's reduced motion preferences with simplified variants.
 *
 * @example
 * ```tsx
 * // AI message (from left)
 * <MessageAnimator direction="left" id="msg-1">
 *   <div className="message ai">Hello!</div>
 * </MessageAnimator>
 *
 * // User message (from right)
 * <MessageAnimator direction="right" id="msg-2" delay={0.1}>
 *   <div className="message user">Hi there!</div>
 * </MessageAnimator>
 * ```
 */

import { ReactNode, useEffect, useState } from 'react';
import { motion } from 'framer-motion';
import { useReducedMotion } from '@/lib/animations';

export interface MessageAnimatorProps {
  /**
   * Child elements to animate
   */
  children: ReactNode;

  /**
   * Direction of slide animation
   * - 'left': AI messages (slide from left, x: -20)
   * - 'right': User messages (slide from right, x: 20)
   */
  direction: 'left' | 'right';

  /**
   * Animation delay in seconds
   * @default 0
   */
  delay?: number;

  /**
   * Unique message identifier
   */
  id: string;
}

/**
 * MessageAnimator component for chat message animations
 */
export function MessageAnimator({
  children,
  direction,
  delay = 0,
  id,
}: MessageAnimatorProps) {
  const shouldReduceMotion = useReducedMotion();
  const [animationComplete, setAnimationComplete] = useState(false);

  // Determine initial x position based on direction
  const initialX = direction === 'left' ? -20 : 20;

  // Create variants based on reduced motion preference
  const variants = shouldReduceMotion
    ? {
        // Simplified variants for reduced motion
        initial: { opacity: 1, x: 0 },
        animate: { opacity: 1, x: 0 },
      }
    : {
        // Full animation variants
        initial: { opacity: 0, x: initialX },
        animate: {
          opacity: 1,
          x: 0,
        },
      };

  // Transition configuration
  const transition = shouldReduceMotion
    ? { duration: 0 }
    : {
        type: 'spring' as const,
        stiffness: 500,
        damping: 30,
        delay,
      };

  // Set animation complete after delay + animation duration
  useEffect(() => {
    if (!shouldReduceMotion) {
      const animationDuration = 300; // Spring animation approx duration
      const totalDelay = (delay * 1000) + animationDuration;
      const timer = setTimeout(() => {
        setAnimationComplete(true);
      }, totalDelay);
      return () => clearTimeout(timer);
    } else {
      // Immediate completion for reduced motion
      setAnimationComplete(true);
    }
  }, [delay, shouldReduceMotion]);

  return (
    <motion.div
      variants={variants}
      initial="initial"
      animate="animate"
      transition={transition}
      data-message-id={id}
      data-animation-complete={animationComplete.toString()}
    >
      {children}
    </motion.div>
  );
}
