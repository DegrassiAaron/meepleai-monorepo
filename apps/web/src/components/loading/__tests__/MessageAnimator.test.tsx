/**
 * Tests for MessageAnimator component
 * Wrapper for animating chat messages with direction-specific slide-in effects
 */

import { render, screen } from '@testing-library/react';
import { MessageAnimator } from '../MessageAnimator';

// Mock useReducedMotion hook
const mockUseReducedMotion = jest.fn();
jest.mock('@/lib/animations', () => ({
  ...jest.requireActual('@/lib/animations'),
  useReducedMotion: () => mockUseReducedMotion(),
}));

// Mock framer-motion with props tracking
const mockMotionDiv = jest.fn(({ children, ...props }: any) => (
  <div {...props}>{children}</div>
));

jest.mock('framer-motion', () => ({
  motion: {
    div: (props: any) => mockMotionDiv(props),
  },
}));

describe('MessageAnimator', () => {
  beforeEach(() => {
    mockUseReducedMotion.mockReturnValue(false);
    mockMotionDiv.mockClear();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('Children wrapping', () => {
    it('should wrap children correctly', () => {
      render(
        <MessageAnimator direction="left" id="msg-1">
          <div>Test message content</div>
        </MessageAnimator>
      );
      expect(screen.getByText('Test message content')).toBeInTheDocument();
    });

    it('should wrap multiple children', () => {
      render(
        <MessageAnimator direction="right" id="msg-2">
          <div>First child</div>
          <div>Second child</div>
        </MessageAnimator>
      );
      expect(screen.getByText('First child')).toBeInTheDocument();
      expect(screen.getByText('Second child')).toBeInTheDocument();
    });

    it('should wrap text nodes', () => {
      render(
        <MessageAnimator direction="left" id="msg-3">
          Plain text message
        </MessageAnimator>
      );
      expect(screen.getByText('Plain text message')).toBeInTheDocument();
    });
  });

  describe('Direction-specific animation', () => {
    it('should apply left direction animation (x: -20)', () => {
      render(
        <MessageAnimator direction="left" id="msg-4">
          <div>AI Message</div>
        </MessageAnimator>
      );
      expect(mockMotionDiv).toHaveBeenCalled();
      const lastCall = mockMotionDiv.mock.calls[mockMotionDiv.mock.calls.length - 1][0];
      expect(lastCall.variants.initial.x).toBe(-20);
      expect(lastCall.variants.animate.x).toBe(0);
    });

    it('should apply right direction animation (x: 20)', () => {
      render(
        <MessageAnimator direction="right" id="msg-5">
          <div>User Message</div>
        </MessageAnimator>
      );
      expect(mockMotionDiv).toHaveBeenCalled();
      const lastCall = mockMotionDiv.mock.calls[mockMotionDiv.mock.calls.length - 1][0];
      expect(lastCall.variants.initial.x).toBe(20);
      expect(lastCall.variants.animate.x).toBe(0);
    });

    it('should include opacity animation for both directions', () => {
      const { rerender } = render(
        <MessageAnimator direction="left" id="msg-6">
          <div>Message</div>
        </MessageAnimator>
      );
      let lastCall = mockMotionDiv.mock.calls[mockMotionDiv.mock.calls.length - 1][0];
      expect(lastCall.variants.initial.opacity).toBe(0);
      expect(lastCall.variants.animate.opacity).toBe(1);

      mockMotionDiv.mockClear();
      rerender(
        <MessageAnimator direction="right" id="msg-7">
          <div>Message</div>
        </MessageAnimator>
      );
      lastCall = mockMotionDiv.mock.calls[mockMotionDiv.mock.calls.length - 1][0];
      expect(lastCall.variants.initial.opacity).toBe(0);
      expect(lastCall.variants.animate.opacity).toBe(1);
    });
  });

  describe('Delay prop', () => {
    it('should respect delay prop', () => {
      render(
        <MessageAnimator direction="left" id="msg-8" delay={0.5}>
          <div>Delayed Message</div>
        </MessageAnimator>
      );
      const lastCall = mockMotionDiv.mock.calls[mockMotionDiv.mock.calls.length - 1][0];
      expect(lastCall.transition.delay).toBe(0.5);
    });

    it('should default to 0 delay when not provided', () => {
      render(
        <MessageAnimator direction="left" id="msg-9">
          <div>Message</div>
        </MessageAnimator>
      );
      const lastCall = mockMotionDiv.mock.calls[mockMotionDiv.mock.calls.length - 1][0];
      expect(lastCall.transition.delay).toBe(0);
    });
  });

  describe('Reduced motion', () => {
    it('should respect prefers-reduced-motion', () => {
      mockUseReducedMotion.mockReturnValue(true);
      render(
        <MessageAnimator direction="left" id="msg-10">
          <div>Message</div>
        </MessageAnimator>
      );
      const lastCall = mockMotionDiv.mock.calls[mockMotionDiv.mock.calls.length - 1][0];
      // Simplified variants when reduced motion is true
      expect(lastCall.variants.initial.x).toBe(0);
      expect(lastCall.variants.animate.x).toBe(0);
      expect(lastCall.variants.initial.opacity).toBe(1);
      expect(lastCall.variants.animate.opacity).toBe(1);
    });
  });

  describe('Data attributes', () => {
    it('should set data-message-id attribute', () => {
      const { container } = render(
        <MessageAnimator direction="left" id="msg-11">
          <div>Message</div>
        </MessageAnimator>
      );
      const wrapper = container.firstChild;
      expect(wrapper).toHaveAttribute('data-message-id', 'msg-11');
    });

    it('should set data-animation-complete attribute', () => {
      const { container } = render(
        <MessageAnimator direction="right" id="msg-12">
          <div>Message</div>
        </MessageAnimator>
      );
      const wrapper = container.firstChild;
      expect(wrapper).toHaveAttribute('data-animation-complete', 'false');
    });

    it('should handle different message IDs', () => {
      const { container: container1 } = render(
        <MessageAnimator direction="left" id="unique-id-123">
          <div>Message 1</div>
        </MessageAnimator>
      );
      expect(container1.firstChild).toHaveAttribute('data-message-id', 'unique-id-123');

      const { container: container2 } = render(
        <MessageAnimator direction="right" id="another-id-456">
          <div>Message 2</div>
        </MessageAnimator>
      );
      expect(container2.firstChild).toHaveAttribute('data-message-id', 'another-id-456');
    });
  });

  describe('Snapshot tests', () => {
    it('should match snapshot for left direction', () => {
      const { container } = render(
        <MessageAnimator direction="left" id="msg-13">
          <div>AI Message</div>
        </MessageAnimator>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for right direction', () => {
      const { container } = render(
        <MessageAnimator direction="right" id="msg-14">
          <div>User Message</div>
        </MessageAnimator>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
