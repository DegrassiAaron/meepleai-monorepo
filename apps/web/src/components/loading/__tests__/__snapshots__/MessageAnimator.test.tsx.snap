// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`MessageAnimator Snapshot tests should match snapshot for left direction 1`] = `
<div
  animate="animate"
  data-animation-complete="false"
  data-message-id="msg-13"
  initial="initial"
  transition="[object Object]"
  variants="[object Object]"
>
  <div>
    AI Message
  </div>
</div>
`;

exports[`MessageAnimator Snapshot tests should match snapshot for right direction 1`] = `
<div
  animate="animate"
  data-animation-complete="false"
  data-message-id="msg-14"
  initial="initial"
  transition="[object Object]"
  variants="[object Object]"
>
  <div>
    User Message
  </div>
</div>
`;
